### PR TITLE
SFTPGo 2.2.3 (new formula)

### PR DESCRIPTION
**Pull request by drakkan**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/sftpgo.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  depends_on "go" => :build
+
+  def install
+    system "go", "build", "-ldflags", "-s -w", "-trimpath", "-o", "sftpgo"
+    inreplace "sftpgo.json" do |s|
+      s.gsub! "\"users_base_dir\": \"\"", "\"users_base_dir\": \"#{var}/sftpgo/data\""
+      s.gsub! "\"backups_path\": \"backups\"", "\"backups_path\": \"#{var}/sftpgo/backups\""
+      s.gsub! "\"credentials_path\": \"credentials\"", "\"credentials_path\": \"#{var}/sftpgo/credentials\""
+    end
+    bin.install "sftpgo"
```
Comment (carlocab) [inline Formula/sftpgo.rb:17]:
```suggestion
    system "go", "build", *std_go_args(ldflags: "-s -w")
    inreplace "sftpgo.json" do |s|
      s.gsub! "\"users_base_dir\": \"\"", "\"users_base_dir\": \"#{var}/sftpgo/data\""
      s.gsub! "\"backups_path\": \"backups\"", "\"backups_path\": \"#{var}/sftpgo/backups\""
      s.gsub! "\"credentials_path\": \"credentials\"", "\"credentials_path\": \"#{var}/sftpgo/credentials\""
    end
```

File: Formula/sftpgo.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+      #{pkgetc}/sftpgo.json
+
+      Getting started guide:
+
+      https://github.com/drakkan/sftpgo/blob/main/docs/howto/getting-started.md
+
+      Step-to-step tutorials:
+
+      https://github.com/drakkan/sftpgo/tree/main/docs/howto
```
Comment (carlocab) [inline Formula/sftpgo.rb:51]:
```suggestion
      Default data location:

      #{var}/sftpgo

      Configuration file location:

      #{pkgetc}/sftpgo.json
```
Most of this belongs in upstream documentation.

File: Formula/sftpgo.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    EOS
+  end
+
+  plist_options startup: true
+
+  def plist
```
Comment (carlocab) [inline Formula/sftpgo.rb:57]: Can we use a `service` block instead?

File: Formula/sftpgo.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        <true/>
+        <key>ThrottleInterval</key>
+        <integer>10</integer>
+      </dict>
+      </plist>
+    EOS
+  end
+
+  test do
+    assert_match version.to_s, shell_output("#{bin}/sftpgo -v")
```
Comment (carlocab) [inline Formula/sftpgo.rb:99]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

Comment (drakkan) [reply Formula/sftpgo.rb:78]: > Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.
> 
> > We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests.
> 
> Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

Thanks for your review, I'm a bit unsure here, this a daemon, do you want that I start the service and check that is running after an arbitrary sleep time?

File: Formula/sftpgo.rb (hunk 3 above)
Comment (drakkan) [reply Formula/sftpgo.rb:36]: I'm not a macOS expert, so this question could be trivial. I would like to run the service as non-root user, so in the service definition I have set

```
<key>UserName</key>
<string>#{ENV["USER"]}</string>
```

how can I do this in a service block? Thanks

Comment (SMillerDev) [reply Formula/sftpgo.rb:36]: That doesn't actually work in `brew services`, but `brew services` runs as non-root user by default anyway.

File: Formula/sftpgo.rb (hunk 4 above)
Comment (SMillerDev) [reply Formula/sftpgo.rb:78]: Yeah, that sounds good

File: Formula/sftpgo.rb (hunk 3 above)
Comment (drakkan) [reply Formula/sftpgo.rb]: strange, maybe there is something wrong in my local installation then

```
sudo brew services info sftpgo
sftpgo (homebrew.mxcl.sftpgo)
Running: ✔
Loaded: ✔
Schedulable: ✘
User: root
PID: 52423
```

this is the output I get after installing the just pushed version. As said before I don't know anything about macOS, this is the first time I use homebrew :smile:

File: Formula/sftpgo.rb (hunk 4 above)
Comment (drakkan) [reply Formula/sftpgo.rb]: > Yeah, that sounds good

let me know if this is ok now. The test now checks the http service using curl and the sftp service by doing a keyscan

File: Formula/sftpgo.rb [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    ENV["SFTPGO_SFTPD__BINDINGS__0__ADDRESS"] = "127.0.0.1"
+    ENV["SFTPGO_SFTPD__BINDINGS__0__PORT"] = sftp_port.to_s
+    ENV["SFTPGO_LOG_FILE_PATH"] = "#{testpath}/sftpgo.log"
+    pid = fork do
+      exec bin/"sftpgo", "serve", "--config-file", "#{pkgetc}/sftpgo.json"
+    end
+
+    sleep 10
+    assert_match expected_output, shell_output("curl -s 127.0.0.1:#{http_port}/healthz")
+    shell_output("ssh-keyscan -p #{sftp_port} 127.0.0.1")
```
Comment (carlocab) [inline Formula/sftpgo.rb:64]: If you just need to run a command to check that it succeeds, use `system`. `shell_output` is for when you need to capture the output of the command.

Comment (drakkan) [reply Formula/sftpgo.rb]: done, thanks

File: Formula/sftpgo.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  desc "Fully featured SFTP server with optional HTTP/S, FTP/S and WebDAV support"
+  homepage "https://github.com/drakkan/sftpgo"
+  url "https://github.com/drakkan/sftpgo/releases/download/v2.2.3/sftpgo_v2.2.3_src_with_deps.tar.xz"
+  sha256 "6c8676725e86ee3f6ad46a340a84f0da37cab8b6ea7b6aee86b2b96ba5e6671a"
+  license "AGPL-3.0-only"
+
+  depends_on "go" => :build
+
+  def install
+    system "go", "build", "-trimpath", *std_go_args(ldflags: "-s -w"), "-o", "sftpgo"
```
Comment (bayandin) [inline Formula/sftpgo.rb:11]: `-trimpath` — is the one of  std_go_args, `-o` by default points to `#{bin}/sftpgo`

```suggestion
    system "go", "build", *std_go_args(ldflags: "-s -w")
```

File: Formula/sftpgo.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    inreplace "sftpgo.json" do |s|
+      s.gsub! "\"users_base_dir\": \"\"", "\"users_base_dir\": \"#{var}/sftpgo/data\""
+      s.gsub! "\"backups_path\": \"backups\"", "\"backups_path\": \"#{var}/sftpgo/backups\""
+      s.gsub! "\"credentials_path\": \"credentials\"", "\"credentials_path\": \"#{var}/sftpgo/credentials\""
+      s.gsub! "\"templates_path\": \"templates\"", "\"templates_path\": \"#{opt_pkgshare}/templates\""
+      s.gsub! "\"static_files_path\": \"static\"", "\"static_files_path\": \"#{opt_pkgshare}/static\""
+      s.gsub! "\"openapi_path\": \"openapi\"", "\"openapi_path\": \"#{opt_pkgshare}/openapi\""
+    end
+    bin.install "sftpgo"
```
Comment (bayandin) [inline Formula/sftpgo.rb:20]:
```suggestion
```

File: Formula/sftpgo.rb [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      Default data location:
+
+      #{var}/sftpgo
+
+      Configuration file location:
+
+      #{pkgetc}/sftpgo.json
+
+      If the SFTPGo service does not start, make sure that TCP ports 2022 and 8080 are not used by other services
+      or change the SFTPGo configuration to suit your needs.
```
Comment (bayandin) [inline Formula/sftpgo.rb:37]: This is a general thing and not related to the homebrew installation.

```suggestion
```

File: Formula/sftpgo.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    (fish_completion/"sftpgo.fish").write Utils.safe_popen_read(bin/"sftpgo", "gen", "completion", "fish")
+
+    inreplace "sftpgo.json" do |s|
+      s.gsub! "\"users_base_dir\": \"\"", "\"users_base_dir\": \"#{var}/sftpgo/data\""
+      s.gsub! "\"backups_path\": \"backups\"", "\"backups_path\": \"#{var}/sftpgo/backups\""
+      s.gsub! "\"credentials_path\": \"credentials\"", "\"credentials_path\": \"#{var}/sftpgo/credentials\""
+      s.gsub! "\"templates_path\": \"templates\"", "\"templates_path\": \"#{opt_pkgshare}/templates\""
+      s.gsub! "\"static_files_path\": \"static\"", "\"static_files_path\": \"#{opt_pkgshare}/static\""
+      s.gsub! "\"openapi_path\": \"openapi\"", "\"openapi_path\": \"#{opt_pkgshare}/openapi\""
+    end
```
Comment (SMillerDev) [inline Formula/sftpgo.rb:26]: Could you make an upstream issue to make these configurable at build time? It doesn't have to block merging, but this isn't the most future proof way to set these

**Review by SMillerDev — DISMISSED** [review]

File: Formula/sftpgo.rb (hunk 9 above)
Comment (drakkan) [reply Formula/sftpgo.rb]: > Could you make an upstream issue to make these configurable at build time? It doesn't have to block merging, but this isn't the most future proof way to set these

@SMillerDev , I'm the upstream :smile:.

I removed some replaces. In the future I can add some automatic search paths for `templates_path`, `static_files_path` and `static_files_path`. The only one that must stay is the `users_base_dir`. This dir defines the global default path for user's home dir (where the user data is stored) and so end users should be able to change it.

I'll add these changes for v2.3.0 (hopefully no other 2.2.x releases will happen).

I'm concerned regarding the user running the service, based on the info [here](https://www.launchd.info/), a global daemon runs on behalf of root or the user specified with the key `UserName`. So I cannot have the SFTPGo service to automatically start on boot as non-root user without defining a custom plist.

Comment (drakkan) [reply Formula/sftpgo.rb]: > Could you make an upstream issue to make these configurable at build time? It doesn't have to block merging, but this isn't the most future proof way to set these

the directories in the shared data paths are now automatically discovered

https://github.com/drakkan/sftpgo/commit/80da2dc722ff50f64a73ee5c1955bfd8de89dc27#diff-a61d5e39f45fb5ed2dc1ef71fa708c8db480e9a396adb8a5d9d17b8e6ba1e186R352

so in v2.3.0 we can remove the replace for `static`, `templates` and `openapi` from the formula, only `users_base_dir` will remain

**Comment by drakkan** [comment]
Is there anything more I can do on my side? Thanks

**Review by chenrui333 — APPROVED** [review]

**Review by chenrui333 — DISMISSED** [review]

File: Formula/sftpgo.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+  url "https://github.com/drakkan/sftpgo/releases/download/v2.2.3/sftpgo_v2.2.3_src_with_deps.tar.xz"
+  sha256 "6c8676725e86ee3f6ad46a340a84f0da37cab8b6ea7b6aee86b2b96ba5e6671a"
+  license "AGPL-3.0-only"
+
+  depends_on "go" => :build
+
+  def install
+    system "go", "build", *std_go_args(ldflags: "-s -w")
+    system bin/"sftpgo", "gen", "man", "-d", "man"
+    man1.install Dir["man/*.1"]
```
Comment (carlocab) [inline Formula/sftpgo.rb:13]:
```suggestion
    system bin/"sftpgo", "gen", "man", "-d", man1
```
Does this work?

Comment (drakkan) [reply Formula/sftpgo.rb]: > Does this work?

Yes, it works, thanks

**Review by carlocab — APPROVED** [review]
Thanks!

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
@drakkan, thanks for your contribution to Homebrew! :tada: 🥇

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

**Comment by drakkan** [comment]
> @drakkan, thanks for your contribution to Homebrew! tada 1st_place_medal
> 
> Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

Thanks to you guys!

**Pull request closed without merging**